### ISSUE TITLE
Costbo (SA: RET13) — compliance checks

### ISSUE THREAD

**Issue by sandeepshahi** [ONDC-Official/v1.2.0-logs#1376]
## Flow 1

**/on_search**

- invalid static terms provided by BNP; SNP must respond with a NACK
- entire address should not be provided in /locations/address/locality
- radius of 3000km is invalid; use realistic value as mentioned in the previous logs feedback

**/on_search (inc)**

- formatting (such as "\n") should not be present in the description
- time/timestamp can't be future dated; not in sync with the full catalog either

## Flow 2

**/on_init**

- provider_tax_number must be provided in bpp_terms

**/on_status**

- Packed: updated_at timestamp can't be even before the Pending state (/on_status) is called
- Out-for-delivery:
  - message_id should be unique
  - pickup timestamp is incorrect; can be even before the Agent is assigned

## Flow 5

**/on_cancel (RTO)**

- start and end location must be present in RTO fulfillment
- all the adjusted line items in quote should be provided in quote_trail (delivery, packing etc)

**/on_status (RTO-Disposed)**

- order/updated_at timestamp should be updated correctly as per the change in order object

## Flow 6

**/on_update (return)**

- refund (part cancel) info received must be appended in settlement details in /payment

@AnanthBo

**Comment by AnanthBo** [comment]
@sandeepshahi ,   Have one question on the Flow 5:,  ON_cancel. comment.      If the seller wants to charge for  foward delivery, misc and Packaging,.   Then he does not need to add in quote trail ?    Leave the positive value in quote and buyer apps will bear that cost as RTO with valid reason, is failure at buyerapp side. .  Please let us know. .      If this is acceptable,  we dont need to add  delivery & packaging to quote_trail.

**Comment by sandeepshahi** [comment]
@AnanthBo, Yes, if the RTO reason id is valid where the cost is attributed to BNP, then it's acceptable. But make sure, any adjustment in quote is reflected in the quote_trail.

**Comment by AnanthBo** [comment]
@sandeepshahi ,  Also  On the comment 1: Does it mean that empty string from buyer is not acceptable.  ?    quote: "  invalid static terms provided by BNP; SNP must respond with a NACK."       

 At  onsearch > tags> code: bap_terms > {
                            "code": "static_terms",
                            "value": ""            <<<< Empty string  is not acceptable. & should be NACKed ? 
                        },

**Comment by AnanthBo** [comment]
@sandeepshahi ,  could you please comment on the above.  quote: " invalid static terms provided by BNP; SNP must respond with a NACK."

At onsearch > tags> code: bap_terms > {
"code": "static_terms",
"value": "" <<<< Empty string is not acceptable. & should be NACKed ?
},

**Comment by sandeepshahi** [comment]
@AnanthBo, static terms are not yet enabled by the NPs. Once enabled, the counterparty NPs (SNP) have to accept the terms in /on_search after going through the static terms. Can you explain why are these static terms accepted here?

**Comment by AnanthBo** [comment]
@sandeepshahi ,   could i understand the  comment is.   As of today,  If any one sends  static_terms. (unsupported feature),  should we NACK it  instead of ignoring it .    Ignoring has value, as we could get more orders.  would that be accepted. ?  Could we chat over phone few mins.

**Comment by AnanthBo** [comment]
@sandeepshahi    RET13 logs have been submitted with the minor corrections commented.  PR#1447
https://github.com/ONDC-Official/v1.2.0-logs/pull/1447

**Comment by AnanthBo** [comment]
@sandeepshahi ,   Since we are already present on RET10 & RET15.  and we have fixed the requested information.  Could you please approve RET13.  Also Search is where we will have difference.

**Comment by AnanthBo** [comment]
@sandeepshahi   Certification team has given a go ahead (pass)  for RET13 .    Could you approve as well.

**Comment by Jagannath-wits** [comment]
### Flow 2 and 3

**on_status_picked:**
- **tmpstmp2:** : Order's updated_at timestamp should be less than or equal to the context timestamp for the /on_status_picked API.
- **updatedAtTime:**: Order's updated_at timestamp can't be future-dated (should match context/timestamp).

### Flow 5

**invldQuoteTrailPrices:**
- Quote trail price and item quote price sum for on_cancel should be equal to the price as in on_confirm.

### Flow 6-b and 6-c 

Please provide complete logs for Flow 6-b and 6-c ( on_update_interim, on_update_approval, on_update_picked, update_settlement are missing ) 

@AnanthBo @sandeepshahi

**Comment by AnanthBo** [comment]
@jagannath-wits.  @sandeepshahi 

Flow 2 and Flow 3. - updated_at timer fixes.  we have reupdated the logs directly.  with Flow 2. 

Flow 5  -  We discussed with Sandeep,  in our case,  Because we are charging the RTO cost on the buyer App,  which has given a wrong consumer address.   quote trail  overall will not match.  but item level it will match. 

Flow 6b  -logs are added for  liquidation .    6c  -   Return approval flow is must for Fashion only.   We have a exception for this.  We will manually handle this flow in interim.   As of now 0% return cases for us in live. 

Lets connect over phone call.  if you need more clarification.

**Comment by sandeepshahi** [comment]
@AnanthBo, can you please share the latest PR.

For Flow 5, any adjustment in quote should be reflected in quote_trail. Therefore, the calculation must be correct as mentioned by @Jagannath-wits 

For Flow 6b, can you share the approval email with us.

**Comment by AnanthBo** [comment]
@sandeepshahi  @Jagannath-wits 

I do not have email or phone num  of jagannath...Hence attaching the older screenshots  of email &  Message on the RTO quote_trail.  above.  (We will charge the Buyerapp for the RTO costs because buyer address issue,).  with this overall quote will not match.  (but item level it will match). 

![WhatsApp Image 2024-05-02 at 18 35 11](https://github.com/ONDC-Official/v1.2.0-logs/assets/133037547/84768493-ceef-4e09-8b6b-2b3a6cee7fa2)
![WhatsApp Image 2024-05-02 at 18 32 50](https://github.com/ONDC-Official/v1.2.0-logs/assets/133037547/de7ec53a-96ca-49c7-97aa-3f2374a05820)

**Comment by AnanthBo** [comment]
@sandeepshahi  @Jagannath-wits   Please update us at the earliest.

**Comment by sahil-ondc** [comment]
## Flow 1

### on_search

- /message/catalog/bpp/descriptor/images should be publically accessible
- /message/catalog/bpp/providers/items/descriptor/code should be EAN
- /message/catalog/bpp/providers/items/descriptor short_desc and long_desc not matching with item.quanitity.unitized.measure
- Varaints should be provided
- in order timing tags, day_from and day_to should remain same as days provided in /message/catalog/bpp/providers/locations/time
- in order timing tags, time_from and time_to should remain same as range provided in /message/catalog/bpp/providers/locations/time

## Flow 2

### on_init

- /message/order/tags/bpp_terms should have property np_type

### on_confirm

- pickup and delivery ranges can't overlap

### on_status (pending)

- /message/order/fulfillment/end/location/gps can't change (on_confirm: 12.920255557348954,77.69561290740968, on_status: 12.920255,77.695612)
- tracking information should be provided in /message/order/fulfillment/tags

### on_status (picked)

- Convenience Fee, Packing charges missing in invoice, but provided in quote

@AnanthBo

**Comment by Jagannath-wits** [comment]
#### Flow 1: on_search_full_catalog_refresh

- `"schemaErr0"`: `/message/catalog/bpp~1fulfillments/2/type` must be equal to one of the allowed values (Delivery, Self-Pickup).

#### Flow 2: select

- `"tmpstmp"`: Timestamp for `/on_search` API cannot be greater than or equal to `/select` API.
- `"invalidItemId[5b26ad1746f3492a88a166f504d25ec4]"`: Invalid item found in `/select` with ID: 5b26ad1746f3492a88a166f504d25ec4.
- `"itemProvider[1]"`: Item with ID 5b26ad1746f3492a88a166f504d25ec4 is not available for provider with ID 7bf9b561199f4a898b3f8a3cef91b6e6.

#### Flow 3: select_out_of_stock
- same as above

#### Flow 4: on_cancel

- `"precancelState.updatedAt"`: precancel_state.updated_at of on_cancel is not equal with the on_confirm order.updated_at.

#### Flow 5: on_cancel

- `"invldQuoteTrailPrices"`: Quote trail price and item quote price sum for on_cancel should be equal to the price as in on_confirm.
On_status_rto_disposed : end.timestamp missing in RTO fulfillment object


#### Flow 6
- Missing update_settlement of reverse qc and calls after on_update_return_initiated 

@AnanthBo Please Fix the above issues; No need to re-submit the logs.
